asab/web/tenant/service.py

```diff
@@ -81,15 +65,37 @@ def add_web_api(self, web_container):
 
 
 	async def _update_tenants(self, message_name=None):
-		async with aiohttp.ClientSession() as session:
-			async with session.get(self.TenantUrl) as resp:
-				if resp.status == 200:
-					tenants_list = await resp.json()
-					for tenant in tenants_list:
-						if isinstance(tenant, str):
-							self.Tenants.add(tenant)
-						elif isinstance(tenant, dict) and "_id" in tenant:
-							# TODO: REMOVE?
-							self.Tenants.add(tenant["_id"])
-						else:
-							L.warning("Unknown tenant format: {}".format(tenant))
+		new_tenants = set()
+
+		if len(self.TenantUrl) > 0:
+			async with aiohttp.ClientSession() as session:
+				async with session.get(self.TenantUrl) as resp:
+					if resp.status == 200:
+						external_tenants = await resp.json()
+					else:
+						L.warning("Failed to load tenants.", struct_data={"url": self.TenantUrl})
+						return
+
+			new_tenants.update(external_tenants)
+
+		if len(self._StaticTenants) > 0:
+			new_tenants.update(self._StaticTenants)
+
+		if self.Tenants != new_tenants:
+			self.App.PubSub.publish("Tenants.changed!")
```

Comment:
Maybe "Tenants.change!", same as "Library.change!"?

Reply:
i think "changed" is more accurate (or less ambiguous).

but true, let's stay consistent. i will rename it.